### PR TITLE
Update the build system & QT5 port

### PR DESCRIPTION
**Pull request by ghost**
# Current status

### What I'm doing now

Building the system with wayland and QT5 and trying to port PINN to QT5.

### What I modified

- [x] Added PINN and rebootp as packages
- [x] Added qt5 port of arora to buildroot
- [x] Update version of Raspberry Pi packages
- [x] GCC update to 9.x.x
- [x] Binutils update to 2.33.1
- [x] Use improved arora
- [x] Modified the rpi-firmware package to support all RPis at once
- [x] Add `update-all` and `update-wifi-firmware` to BUILDME.sh

### Builds on
- [ ] Ubuntu 20.04
- [ ] Ubuntu 14.04
- [ ] Debian 10
- [ ] Debian 9

### Works on

- [ ] Raspberry Pi Zero
- [ ] Raspberry Pi Zero W
- [ ] Raspberry Pi Zero WH
- [ ] Raspberry Pi A
- [ ] Raspberry Pi A+
- [ ] Raspberry Pi B
- [ ] Raspberry Pi 2B
- [ ] Raspberry Pi 2B rev. 1.2
- [ ] Raspberry Pi 3A
- [ ] Raspberry Pi 3B
- [ ] Raspberry Pi 3B+
- [ ] Raspberry Pi 4B 

### Advantages of the new buildroot

- All packages updated, possible bug fixes
- wpa3 is now supported in wpa_supplicant
- pinn will use qt5
- arora will render HTML5 sites better
- Pinn can be built on modern systems

### Other information 

- I used an existing QT5 version of Arora and added a new feature.
- The buildroot Linux makefiles are only slightly changed to remove a bug, since buildroot is finally able to let the kernel automatically build the dtb.
- The rpi-firmware makefile is changed to install firmware for all Pis (in buildroot 2020.02 you can choose between firmware for either Pi 1-3 and Zero or Pi 4).
- The password is now stored as SHA-256.
- QT5 can't use windows like QT4 since QWS is missing in QT5. I'm currently using wayland instead.
- The new version of PINN requires at least QT 5.12
- I added a guide on how to upgrade buildroot.

### The latest test build

Not finished yet...

### CODE REVIEW

**Comment by lurch** [comment]
In #302 you confirmed that "It's only a display bug in QT Linguist." - does that mean that all the changes to the `width` fields in the `.ui` files aren't needed?
IMHO the fewer changes each PR contains, the easier it is to review.

**Comment by ghost** [comment]
> In #302 you confirmed that "It's only a display bug in QT Linguist." - does that mean that all the changes to the `width` fields in the `.ui` files aren't needed?

No, It works after changing width.

**Comment by RaspberryPiFan** [comment]
I think it works, but I turned off my computer at night, so I'm testing today.

**Comment by RaspberryPiFan** [comment]
I fixed much bugs already, it's now building again. Actually, `make menuconfig` doesn't works.

**Comment by bovirus** [comment]
@RaspberryPiFan 

I suggest to divide the PR from current language file update and system updates.

I will load the current updated language file and in a separate PR proposal for the system update.

**Comment by procount** [comment]
@RaspberryPiFan - I'm not sure that creating a new "buildroot-new" folder is the way to go.
In the past, buildroot has been advanced without changing the name of this folder.
Plus, as you are finding, you are having to make changes to other scripts like "make_arch_linux-menuconfig.sh" to take account of this change.

I would prefer that any new buildroot version follows the existing folder structure, then there is less potential for things to break. 
If you must maintain the 2 buildroot folders in parallel whilst developing this, I suggest you rename "buildroot" to "buildroot_old" and rename "buildroot_new" back to "buildroot".

**Comment by RaspberryPiFan** [comment]
> @RaspberryPiFan - I'm not sure that creating a new "buildroot-new" folder is the way to go.
> In the past, buildroot has been advanced without changing the name of this folder.
> Plus, as you are finding, you are having to make changes to other scripts like "make_arch_linux-menuconfig.sh" to take account of this change.
> 
> I would prefer that any new buildroot version follows the existing folder structure, then there is less potential for things to break.
> If you must maintain the 2 buildroot folders in parallel whilst developing this, I suggest you rename "buildroot" to "buildroot_old" and rename "buildroot_new" back to "buildroot".

Okay, I'm going to that.

The new version works now, I'm going to rebuild it in folder "buildroot" as a second test, and then try the final version.

**Comment by RaspberryPiFan** [comment]
This is the new buildroot. Try building on Ubuntu, send errors to me.

**Comment by RaspberryPiFan** [comment]
Oh! It doesn't works? The last build worked.

**Comment by RaspberryPiFan** [comment]
New test: Try this config file:

**Comment by procount** [comment]
Can you remove buildroot_old from the PR? 
I thought you wanted that for your own personal comparison. It makes no sense to include it in github, because we can use git to look at the differences.

**Comment by RaspberryPiFan** [comment]
More fixes coming soon (after a new test)

**Comment by RaspberryPiFan** [comment]
gmp has some problems, I'm creating a fix.

**Comment by RaspberryPiFan** [comment]
gmp fixed

**Comment by RaspberryPiFan** [comment]
rpi-firmware: fixed for RPi4

**Comment by RaspberryPiFan** [comment]
Upload coming soon

**Comment by bovirus** [comment]
@RaspberryPiFan

The German translation is not complete.

Section "OptionsDialog" - Strings translated 41 on 104 - 63 strings untranslated
Section "rerunsetup" - Strings translated 5  on 14 - 9 strings untranslated

For me it's important that you complete the translation.

**Comment by ghost** [comment]
> @RaspberryPiFan
> 
> The German translation is not complete.
> 
> Section "OptionsDialog" - Strings translated 41 on 104 - 63 strings untranslated
> Section "rerunsetup" - Strings translated 5 on 14 - 9 strings untranslated
> 
> For me it's important that you complete the translation.

@lurch want's to fix translations. I can't translate everything because German has another grammer as English.

**Comment by bovirus** [comment]
@RaspberryPiFan 

Ok if you are talking about strings composed by different single sub-string.
The problem about German language it's the same for Italian because the flow of a sentence is different between English and Italian.

For that I asked to the developer to avoid as much as possible the composed strings.

But many untranslated strings in German language are single sentence/strings.

**Comment by procount** [comment]
@bovirus I'm happy for @RaspberryPiFan to complete his porting to the new buildroot first. The translations can wait until @lurch has removed the compound strings, which will make the translation easier anyway.

**Comment by RaspberryPiFan** [comment]
There are some problems with RPi4, but I'm going to upload a testing pinn_lite.zip build with new buildroot in 10 minutes

**Comment by RaspberryPiFan** [comment]
Sorry, i can upload this in 30 minutes

**Comment by RaspberryPiFan** [comment]
Oh, one error at the end, fix upload in 5 minutes.

**Comment by RaspberryPiFan** [comment]
pinn-lite.zip takes more time.

**Comment by RaspberryPiFan** [comment]
rpi_firmware has some problems. Fix is in development.

**Comment by RaspberryPiFan** [comment]
I had to restart building because i fixed a bug.

**Comment by RaspberryPiFan** [comment]
Building is finished, but buildroot doesn't works right.
Link for first build as zip coming soon (untested)

**Comment by RaspberryPiFan** [comment]
Untested pinn (build with new buildroot): https://drive.google.com/drive/folders/18oFwcxcPP-uUYlMAYz3Hz_D8GfOusEdH?usp=sharing

**Comment by RaspberryPiFan** [comment]
On Pi4, I got kernel panic

**Comment by RaspberryPiFan** [comment]
> On Pi4, I got kernel panic

I'm testing a fix for this.

**Comment by RaspberryPiFan** [comment]
The new version is optimized for RPi 2, 3 and 4. Please test it on RPi 1 and zero.

**Comment by RaspberryPiFan** [comment]
GMP fixed. Upload coming soon

**Comment by RaspberryPiFan** [comment]
Some problems with the checksum, fix coming soon.

**Comment by RaspberryPiFan** [comment]
I disabled the checksum function.
I checked all `.hash` files, but I everytimes get the error: `no hash found`.

**Comment by RaspberryPiFan** [comment]
It work's now, I will upload the final version in 7 hours.

**Comment by ghost** [comment]
I don't have enough time now to upload a fix.

**Comment by procount** [comment]
Take your time. 
Updating buildroot is a big change that will require a lot of testing.
I would prefer a version that works properly rather than a quickly delivered one.
Having reverted the previous changes to build under Ubuntu 18.04/10, this issue is not blocking anything, provided PINN is built using Ubuntu 14.04 or equivalent.

**Comment by RaspberryPiFan** [comment]
The system builds now, but it only showsw a rainbow screen at start. I'm fixing it.

**Comment by RaspberryPiFan** [comment]
The problems are missing files: `recovery.cmdline`, `kernelconfig-recovery.armv6`, `kernelconfig-recovery.armv7`, `kernelconfig-recovery.armv7l`

**Comment by RaspberryPiFan** [comment]
I'm fixing it

**Comment by RaspberryPiFan** [comment]
Current status:

### What I'm doing now:

- building & testing


### What I modified in buildroot to work: 

- Re-add QT (removed in 2019.02) and removed it from the list of legacy options.
- Added packages from old buildroot
- Downgrade for gmp (new version isn't working)

### What I changed

- Update for the German translation

### Bug fixes

- [ ] Building PINN takes a long time
- [ ] DTS problems
- [ ] Unused files
- [ ] Git clone takes a long time (fix: re-open this pull request when buildroot works)
- [ ] Problems with recovery.cmdline

**Comment by ghost** [comment]
I will make 2 pull requests (language and buildroot) after finishing buildroot.

**Comment by ghost** [comment]
Are buildroot changes needed for #2?

**Comment by procount** [comment]
> Are buildroot changes needed for #2 

Probably not, but that's a separate issue. 
Each Pull request should fix one issue, which is why Bovirus suggested splitting out your translations from these buildroot changes.

**Comment by ghost** [comment]
> > Are buildroot changes needed for #2
> 
> Probably not, but that's a separate issue.
> Each Pull request should fix one issue, which is why Bovirus suggested splitting out your translations from these buildroot changes.

Yes, I'm going to split my translations from this PR after finishing this.

**Comment by RaspberryPiFan** [comment]
It works with NOOBS now, but there are some issues with PINNs cec support.

**Comment by RaspberryPiFan** [comment]
It shows errors like this:

```
ceclistener.cpp: 77: 41: Error: "CEC_User_Control_PreviousChannel" was not defined in this scope
      { "CEC_User_Control_PreviousChannel" CEC_User_Control_PreviousChannel}
```

**Comment by RaspberryPiFan** [comment]
On my German computer the error is german, this is a translation.

**Comment by RaspberryPiFan** [comment]
I'm testing a fix.
I added some more QT modules and updated ceclistener.cpp and ceclistener.h.

**Comment by RaspberryPiFan** [comment]
Building works now, but you need some manual changes, I'm creating a fix for this.

**Comment by RaspberryPiFan** [comment]
Please test this: https://raw.githubusercontent.com/RaspberryPiFan/pinn-zip/master/pinn-new.zip

It's untested, I'm going to test it tommorow

**Comment by ghost** [comment]
It doesn't works.

**Comment by ghost** [comment]
I'm working on it.

**Comment by procount** [comment]
I admire your persistence

**Comment by RaspberryPiFan** [comment]
Please test this: https://github.com/RaspberryPiFan/pinn-zip/blob/master/pinn-lite.zip?raw=true

**Comment by RaspberryPiFan** [comment]
Broken on Rpi 4

**Comment by ghost** [comment]
The latest buildroot working with PINN is commit 4b1c0855b321807abdcd737057ae7436547b8001, newer versions don't have QT4. For this, PINN needs to support QT5, but I think this version will work on future ubuntu version.

**Comment by ghost** [comment]
@lurch @procount @bovirus @XECDesign @ShadowEO 
Can someone please test the version I uploaded and compile it.
If you wonder why another account made this, I didn't wanted this, it's my 2nd account.

**Comment by ghost** [comment]
2nd account

**Comment by procount** [comment]
@RaspberryPiFan2 - Thanks for persevering with this. 
I look forward to testing this out, but I'm afraid it won't be for several weeks yet as I've got something else on. 
Have you tried building this version at all yet?

**Comment by ghost** [comment]
> @RaspberryPiFan2 - Thanks for persevering with this.
> I look forward to testing this out, but I'm afraid it won't be for several weeks yet as I've got something else on.
> Have you tried building this version at all yet?

No, I couldn't, currently I only can build on my Raspberry Pi (my PC is broken), and it's hard to build some host tools like GMP on it.

**Comment by RaspberryPiFan** [comment]
For someone who wants to help with testing:

- [ ] Build PINN with my buildroot
- [ ] Send error messages to me
- [ ] Test it

If it doesn't works:

- [ ] Build original PINN
- [ ] Compress the target directory in the output **subdirectory of buildroot** in old and new version and send it to me (upload in github repo or cloud)

File: BUILDME.sh [hunk 1]
```diff
@@ -1,6 +1,5 @@
 #!/bin/bash
 # Bash script to rebuild recovery
-
```
Comment (RaspberryPiFan) [inline BUILDME.sh]: Probably useless. Why did I do this?

Comment (RaspberryPiFan) [reply BUILDME.sh]: I'll ignore it

**Comment by ghost** [comment]
I will port 3.3.2 later.

**Comment by ghost** [comment]
I'm still working on it.

**Comment by ghost** [comment]
Replaced everything. .config file later

**Comment by ghost** [comment]
Currently testing everything, since this is QT5, PINN probably won't compile, but I will fix it after I know the exact error messages.

**Comment by ghost** [comment]
I used an existing QT5 arora and improved it, https://github.com/RaspberryPiFan/arora/commit/43000bf2d8b286166234f5ba02cf4219144bf473 (this are not the QT5 changes, but it only works with QT5 since it uses QT5 features, the QT5 port is another commit I didn't make)

**Comment by ghost** [comment]
This is how arora now renders pages, thanks to qt 5 and webkit 602.1. Screenshot generated in Raspbian)
image
image

**Comment by ghost** [comment]
Making progress with QT5, haven't tested it yet, custom background feature is unavailable, I'm currently trying to bring it back for QT5

**Pull request closed without merging**